MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc

```diff
@@ -0,0 +1,137 @@
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "DataFormats/GeometryVector/interface/GlobalVector.h"
+#include "DataFormats/FWLite/interface/Record.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "CLHEP/Geometry/Vector3D.h"
+
+#include "MagneticField/ParametrizedEngine/interface/alpaka/ParabolicParametrizedMagneticField.h"
+
+#include "FWCore/Utilities/interface/stringize.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+
+#include <iostream>
+#include <fstream>
+#include <Eigen/Core>
+
+using namespace edm;
+using namespace std;
+using namespace alpaka;
+using namespace ALPAKA_ACCELERATOR_NAMESPACE;
```

Comment:
I'd guess the source of the compilation problem in https://github.com/cms-sw/cmssw/pull/44360#discussion_r1698831093 lies here. `ALPAKA_ACCELERATOR_NAMESPACE` defines type alias `Platform` that aliases one of Alpaka's concrete Platform types, and `alpaka` defines type alias template `Platform<T>` that aliases the Platform the argument `T` device. This results either an ambiguity, or the compiler attempts to use the `Platform<T>` alias and fails because the argument `T` is missing.

I'd remove at least the `using namespace alpaka;`.

Reply:
Thanks @makortel, that worked.
Now I have this error that I can't understand:

```
[lecriste@lxplus906 src]$ scram b -j
>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Subsystem BigProducts built
>> Building CMSSW version CMSSW_14_1_0_pre0 ----
>> Creating project symlinks
>> Entering Package MagneticField/ParametrizedEngine
>> Compiling alpaka/cuda src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc
>> Compiling alpaka/rocm src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc
>> Compiling alpaka/serial src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc
In file included from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/version.hpp:7,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/alpaka.hpp:11,
                 from src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:1:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h: In instantiation of 'class alpaka::TaskKernelGpuUniformCudaHipRt<alpaka::ApiCudaRt, alpaka::AccGpuUniformCudaHipRt<alpaka::ApiCudaRt, std::integral_constant<long unsigned int, 1>, unsigned int>, std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>':
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/acc/AccGpuUniformCudaHipRt.hpp:265:24:   required from 'static auto alpaka::trait::CreateTaskKernel<alpaka::AccGpuUniformCudaHipRt<TApi, TDim, TIdx>, TWorkDiv, TKernelFnObj, TArgs ...>::createTaskKernel(const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TApi = alpaka::ApiCudaRt; TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TWorkDiv = alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:287:97:   required from 'auto alpaka::createTaskKernel(const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccGpuUniformCudaHipRt<ApiCudaRt, std::integral_constant<long unsigned int, 1>, unsigned int>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:312:46:   required from 'void alpaka::exec(TQueue&, const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccGpuUniformCudaHipRt<ApiCudaRt, std::integral_constant<long unsigned int, 1>, unsigned int>; TQueue = uniform_cuda_hip::detail::QueueUniformCudaHipRt<ApiCudaRt, false>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h:65:63: error: static assertion failed: You should move this files to a .dev.cc file under the alpaka/ subdirectory.
   65 |     static_assert(std::is_same_v<TApi, alpaka::ApiCudaRt> and BOOST_LANG_CUDA,
      |                                                               ^~~~~~~~~~~~~~~
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h:65:63: note: the comparison reduces to '(0 != 0)'
In file included from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/move.h:57,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/exception_ptr.h:43,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/exception:168,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/boost/1.80.0-e4b2127ad40e46379cea2f9087beeda8/include/boost/predef/detail/_exception.h:12,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/boost/1.80.0-e4b2127ad40e46379cea2f9087beeda8/include/boost/predef/library/std/_prefix.h:21,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/boost/1.80.0-e4b2127ad40e46379cea2f9087beeda8/include/boost/predef/library/std.h:12,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/boost/1.80.0-e4b2127ad40e46379cea2f9087beeda8/include/boost/predef/library.h:14,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/boost/1.80.0-e4b2127ad40e46379cea2f9087beeda8/include/boost/predef.h:16,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/core/BoostPredef.hpp:8,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/Op.hpp:7,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/Traits.hpp:7,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/AtomicAtomicRef.hpp:7,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/AtomicCpu.hpp:13,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/acc/AccCpuOmp2Blocks.hpp:8,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/alpaka.hpp:13,
                 from src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:1:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/type_traits: In substitution of 'template<class _Fn, class ... _Args> using invoke_result_t = typename std::invoke_result::type [with _Fn = MagneticFieldKernel; _Args = {const alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>* const&, float* const&, const long unsigned int&}]':
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:237:23:   required from 'void alpaka::detail::CheckFnReturnType<TAcc, TSfinae>::operator()(const TKernelFnObj&, const TArgs& ...) [with TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int}; TAcc = alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>; TSfinae = void]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:265:42:   required from 'auto alpaka::createTaskKernel(const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:312:46:   required from 'void alpaka::exec(TQueue&, const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>; TQueue = QueueGenericThreadsBlocking<DevCpu>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/type_traits:3041:11: error: no type named 'type' in 'struct std::invoke_result<MagneticFieldKernel, const alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>* const&, float* const&, const long unsigned int&>'
 3041 |     using invoke_result_t = typename invoke_result<_Fn, _Args...>::type;
      |           ^~~~~~~~~~~~~~~
In file included from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/mutex:38,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/AtomicStdLibLock.hpp:11,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/AtomicCpu.hpp:14:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple: In instantiation of 'constexpr decltype(auto) std::__apply_impl(_Fn&&, _Tuple&&, index_sequence<_Idx ...>) [with _Fn = const MagneticFieldKernel&; _Tuple = tuple<alpaka::AccCpuSerial<integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int>; long unsigned int ..._Idx = {0, 1, 2, 3}; index_sequence<_Idx ...> = integer_sequence<long unsigned int, 0, 1, 2, 3>]':
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:1863:31:   required from 'constexpr decltype(auto) std::apply(_Fn&&, _Tuple&&) [with _Fn = const MagneticFieldKernel&; _Tuple = tuple<alpaka::AccCpuSerial<integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/TaskKernelCpuSerial.hpp:92:31:   required from 'void alpaka::TaskKernelCpuSerial<TDim, TIdx, TKernelFnObj, TArgs>::operator()() const [with TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/queue/QueueGenericThreadsBlocking.hpp:135:21:   required from 'static void alpaka::trait::Enqueue<alpaka::QueueGenericThreadsBlocking<TDev>, TTask>::enqueue(alpaka::QueueGenericThreadsBlocking<TDev>&, const TTask&) [with TDev = alpaka::DevCpu; TTask = alpaka::TaskKernelCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/queue/Traits.hpp:49:61:   required from 'void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = QueueGenericThreadsBlocking<DevCpu>; TTask = TaskKernelCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:312:16:   required from 'void alpaka::exec(TQueue&, const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>; TQueue = QueueGenericThreadsBlocking<DevCpu>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:1852:27: error: no matching function for call to '__invoke(const MagneticFieldKernel&, alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, std::__tuple_element_t<1, std::tuple<alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int> >, std::__tuple_element_t<2, std::tuple<alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int> >, std::__tuple_element_t<3, std::tuple<alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int> >)'
 1852 |       return std::__invoke(std::forward<_Fn>(__f),
      |              ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
 1853 |                            std::get<_Idx>(std::forward<_Tuple>(__t))...);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/version.hpp:7,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/alpaka.hpp:11,
                 from src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:1:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h: In instantiation of 'class alpaka::TaskKernelGpuUniformCudaHipRt<alpaka::ApiHipRt, alpaka::AccGpuUniformCudaHipRt<alpaka::ApiHipRt, std::integral_constant<long unsigned int, 1>, unsigned int>, std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>':
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/acc/AccGpuUniformCudaHipRt.hpp:265:24:   required from 'static auto alpaka::trait::CreateTaskKernel<alpaka::AccGpuUniformCudaHipRt<TApi, TDim, TIdx>, TWorkDiv, TKernelFnObj, TArgs ...>::createTaskKernel(const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TApi = alpaka::ApiHipRt; TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TWorkDiv = alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:287:97:   required from 'auto alpaka::createTaskKernel(const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccGpuUniformCudaHipRt<ApiHipRt, std::integral_constant<long unsigned int, 1>, unsigned int>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:312:46:   required from 'void alpaka::exec(TQueue&, const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccGpuUniformCudaHipRt<ApiHipRt, std::integral_constant<long unsigned int, 1>, unsigned int>; TQueue = uniform_cuda_hip::detail::QueueUniformCudaHipRt<ApiHipRt, false>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/core/BoostPredef.hpp:27:32: error: static assertion failed: You should move this files to a .dev.cc file under the alpaka/ subdirectory.
   27 | #        define BOOST_LANG_HIP BOOST_VERSION_NUMBER_NOT_AVAILABLE
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h:110:62: note: in expansion of macro 'BOOST_LANG_HIP'
  110 |     static_assert(std::is_same_v<TApi, alpaka::ApiHipRt> and BOOST_LANG_HIP,
      |                                                              ^~~~~~~~~~~~~~
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/core/BoostPredef.hpp:27:32: note: the comparison reduces to '(0 != 0)'
   27 | #        define BOOST_LANG_HIP BOOST_VERSION_NUMBER_NOT_AVAILABLE
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h:110:62: note: in expansion of macro 'BOOST_LANG_HIP'
  110 |     static_assert(std::is_same_v<TApi, alpaka::ApiHipRt> and BOOST_LANG_HIP,
      |                                                              ^~~~~~~~~~~~~~
In file included from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/refwrap.h:38,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/string:51,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/locale_classes.h:40,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/ios_base.h:41,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/ios:42,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/ostream:38,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/iostream:39,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/core/Debug.hpp:9,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/core/Common.hpp:8,
                 from /cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/atomic/Op.hpp:8:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/invoke.h:90:5: note: candidate: 'template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...)'
   90 |     __invoke(_Callable&& __fn, _Args&&... __args)
      |     ^~~~~~~~
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/invoke.h:90:5: note:   template argument deduction/substitution failed:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/invoke.h: In substitution of 'template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...) [with _Callable = const MagneticFieldKernel&; _Args = {alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int}]':
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:1852:27:   required from 'constexpr decltype(auto) std::__apply_impl(_Fn&&, _Tuple&&, index_sequence<_Idx ...>) [with _Fn = const MagneticFieldKernel&; _Tuple = tuple<alpaka::AccCpuSerial<integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int>; long unsigned int ..._Idx = {0, 1, 2, 3}; index_sequence<_Idx ...> = integer_sequence<long unsigned int, 0, 1, 2, 3>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:1863:31:   required from 'constexpr decltype(auto) std::apply(_Fn&&, _Tuple&&) [with _Fn = const MagneticFieldKernel&; _Tuple = tuple<alpaka::AccCpuSerial<integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/TaskKernelCpuSerial.hpp:92:31:   required from 'void alpaka::TaskKernelCpuSerial<TDim, TIdx, TKernelFnObj, TArgs>::operator()() const [with TDim = std::integral_constant<long unsigned int, 1>; TIdx = unsigned int; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/queue/QueueGenericThreadsBlocking.hpp:135:21:   required from 'static void alpaka::trait::Enqueue<alpaka::QueueGenericThreadsBlocking<TDev>, TTask>::enqueue(alpaka::QueueGenericThreadsBlocking<TDev>&, const TTask&) [with TDev = alpaka::DevCpu; TTask = alpaka::TaskKernelCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/queue/Traits.hpp:49:61:   required from 'void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = QueueGenericThreadsBlocking<DevCpu>; TTask = TaskKernelCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int, MagneticFieldKernel, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&>]'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/alpaka/1.1.0-d5100a17ef5e9f41a14b52000051b265/include/alpaka/kernel/Traits.hpp:312:16:   required from 'void alpaka::exec(TQueue&, const TWorkDiv&, const TKernelFnObj&, TArgs&& ...) [with TAcc = AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>; TQueue = QueueGenericThreadsBlocking<DevCpu>; TWorkDiv = WorkDivMembers<std::integral_constant<long unsigned int, 1>, unsigned int>; TKernelFnObj = MagneticFieldKernel; TArgs = {Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, const long unsigned int&}]'
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/invoke.h:90:5: error: no type named 'type' in 'struct std::__invoke_result<const MagneticFieldKernel&, alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 1>, unsigned int>&, Eigen::Matrix<float, 3, 1, 0, 3, 1>*, float*, long unsigned int>'
gmake: *** [config/SCRAM/GMake/Makefile.rules:1824: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticFieldSerialSync/alpaka/testParabolicParametrizedMagneticField.cc.o] Error 1
gmake: *** Waiting for unfinished jobs....
gmake: *** [config/SCRAM/GMake/Makefile.rules:1824: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticFieldCudaAsync/alpaka/testParabolicParametrizedMagneticField.cc.o] Error 1
gmake: *** [config/SCRAM/GMake/Makefile.rules:1824: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticFieldROCmAsync/alpaka/testParabolicParametrizedMagneticField.cc.o] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2

```

Reply:
The cause is in
```
src/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.cc:107:24:   required from here
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/HeterogeneousCore/AlpakaInterface/interface/config.h:65:63: error: static assertion failed: You should move this files to a .dev.cc file under the alpaka/ subdirectory.
   65 |     static_assert(std::is_same_v<TApi, alpaka::ApiCudaRt> and BOOST_LANG_CUDA,
      |              
```
i.e. you should rename the `testParabolicParametrizedMagneticField.cc` to `testParabolicParametrizedMagneticField.dev.cc`.

Reply:
Thanks! Didn't imagine the compilation would fail because of the filename.

Reply:
A `.cc` file is compiled with the host compiler (gcc), whereas a `.dev.cc` file is compiled with the device compilers (gcc / nvcc / hipcc)